### PR TITLE
[INFINITY-502] Add suppress/revive support to DefaultScheduler

### PR DESCRIPTION
**Pull request by mgummelt**
Also makes significant changes to `DefaultRecoveryPlanManager`

### CODE REVIEW

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java [hunk 1]
```diff
@@ -25,95 +23,140 @@
...
-public class DefaultRecoveryPlanManager implements PlanManager {
+public class DefaultRecoveryPlanManager extends ChainedObserver implements PlanManager {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultRecoveryPlanManager.class);
 
     protected volatile Plan plan;
     protected volatile PhaseStrategy phaseStrategy;
     protected volatile Map<UUID, PhaseStrategy> phaseStrategies;
 
     private final StateStore stateStore;
-    private final TaskFailureListener failureListener;
```
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: Is this safe to remove?  It seemed superfluous

**Comment by mgummelt** [comment]
cc @mohitsoni for review

File: src/main/java/org/apache/mesos/curator/CuratorStateStore.java [hunk 2]
```diff
@@ -414,4 +413,33 @@ private String getTasksRootPath() {
...
+            suppressed = serializer.deserialize(bytes, Boolean.class);
+        } catch (IOException e){
+            logger.error("Error converting property " + SUPPRESSED_KEY + " to boolean.", e);
+            throw new StateStoreException(e);
+        }
+
+        return suppressed;
+    }
+
+    public void setSuppressed(final boolean isSuppressed) {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/curator/CuratorStateStore.java]: Is this used to persist `suppressed` flag, so that when a scheduler restarts, it restarts suppressed ? If yes, how  is that implemented ?

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java [hunk 3]
```diff
@@ -18,12 +19,12 @@ public DefaultInstallStrategy(Phase phase) {
...
 
     @Override
-    public Optional<Block> getCurrentBlock() {
+    public Optional<Block> getCurrentBlock(Collection<String> dirtiedAssets) {
         if (interrupted.get()) {
             return Optional.empty();
         } else {
             for (Block block : phase.getBlocks()) {
-                if (!block.isComplete()) {
+                if (!block.isComplete() && !dirtiedAssets.contains(block.getName())) {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java]: `getCurrentBlock()` is a public interface method. `dirtiedAssets` can be `null`. Maybe validate ? `... !CollectionUtils.isNotEmpty(dirtiedAssets) ...`

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultPhase.java [hunk 4]
```diff
@@ -12,6 +12,8 @@
  * Default Phase implementation tracks Blocks both by their UUID and by
  * their ordering. It is an immutable class that can be constructed either
  * directly, or using a fluent style builder.
+ *
+ * A {@DefaultPhase} is an {@Observable} and will forward updates from its {@b}.
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultPhase.java]: `from it's {@b}` ? `Block` ?

File: src/main/java/org/apache/mesos/scheduler/plan/Block.java [hunk 5]
```diff
@@ -11,6 +11,8 @@
  * Defines the interface for a Block of a {@link Phase}. The block is the base unit of a set of
  * tasks to perform, such as launching a Task, updating a Task, or reconciling Mesos state with
  * Framework state. A Block may be in one of four states: PENDING, IN_PROGRESS, COMPLETE, or ERROR.
+ *
+ * A {@Block} is an {@Observable}, and will notify its observers when its state changes.
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/Block.java]: +1 for keeping docs up-to-date.

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultPhaseStrategy.java [hunk 6]
```diff
@@ -63,15 +63,19 @@ private void advancePosition() {
...
+  public Optional<Block> getCurrentBlock(Collection<String> dirtiedAssets) {
     //TODO(nick) the returned Block is not guaranteed to stay in a consistent state.
     //alternative: instead provide a thread-safe wrapper for viewing/changing current block state
     synchronized (this) {
       advancePosition();
       if (shouldStart[currPos]) {
         Block block = phase.getBlocks().get(currPos);
-        logger.info(String.format("Returning block '%s' at position: %d", block.getName(), currPos));
-        return Optional.of(block);
+        if (!dirtiedAssets.contains(block.getName())) {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultPhaseStrategy.java]: Good catch!

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultPlan.java [hunk 7]
```diff
@@ -10,6 +10,8 @@
 /**
  * An ordered list of {@link Phase}s, composed into a {@link Plan}. It may
  * optionally contain a List of errors associated with the phase.
+ *
+ * A {@DefaultPlan} is an {@Observable} and will forward updates from its {@Phase}s.
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultPlan.java]: 👍

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultPlanCoordinator.java [hunk 8]
```diff
@@ -12,8 +13,10 @@
 
 /**
  * Default implementation of PlanCoordinator.
+ *
+ * A {@DefaultPlanCoordinator} is an {@Observable} and will forward updates from its plans.
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultPlanCoordinator.java]: 👍

File: src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java [hunk 9]
```diff
@@ -2,18 +2,22 @@
...
 import org.apache.mesos.SchedulerDriver;
+import org.apache.mesos.scheduler.Observable;
 
 import java.util.Collection;
 import java.util.List;
 
 /**
- * PlanCoordinator job is to perform co-ordination of offers among configured {@code PlanManager}s.
+ * PlanCoordinator's job is to coordinate offers among configured {@code PlanManager}s.
+ * A {@code PlanCoordinator} updates its observers
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java]: `A {@code PlanCoordinator} being an {@link Observable} updates its observers.`

File: src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java [hunk 10]
```diff
@@ -2,18 +2,22 @@
...
  */
-public interface PlanCoordinator {
+public interface PlanCoordinator extends Observable {
     /**
      * Provides offers to each {@code PlanManager} for processing. Keeps tracks of dirtied offers and assets.
      *
      * @return List of OfferID's that were used for scheduling operations.
      */
     Collection<Protos.OfferID> processOffers(final SchedulerDriver driver, final List<Protos.Offer> offers);
+
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java]: Can you please add docs here as well ?

File: src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java [hunk 11]
```diff
@@ -19,12 +16,16 @@ public RandomRecoveryStrategy(Phase phase) {
     }
 
     @Override
-    public Optional<Block> getCurrentBlock() {
+    public Optional<Block> getCurrentBlock(Collection<String> dirtiedAssets) {
         final List<? extends Block> blocks = filterOnlyPendingBlocks(getPhase().getBlocks());
-        if (isInterrupted() || CollectionUtils.isEmpty(blocks)) {
+        if (isInterrupted() || CollectionUtils.isEmpty(blocks) || blocks.size() <= dirtiedAssets.size()) {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java]: `null` check `dirtiedAssets` ?

File: src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java [hunk 12]
```diff
@@ -19,12 +16,16 @@ public RandomRecoveryStrategy(Phase phase) {
...
             return Optional.empty();
         } else {
-            return Optional.of(blocks.get(new Random().nextInt(blocks.size())));
+            List<Block> cleanBlocks = blocks.stream()
+                    .filter(block -> !dirtiedAssets.contains(block.getName()))
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java]: `null` check dirtiedAssets` ?

File: src/main/java/org/apache/mesos/scheduler/plan/ReconciliationStrategy.java [hunk 13]
```diff
@@ -23,9 +24,14 @@ private ReconciliationStrategy(final ReconciliationPhase phase) {
     }
 
     @Override
-    public Optional<Block> getCurrentBlock() {
+    public Optional<Block> getCurrentBlock(Collection<String> dirtiedAssets) {
         // ReconciliationPhases only have a single Block
-        return Optional.of(phase.getBlock(0));
+        for (Block block : phase.getBlocks()) {
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/ReconciliationStrategy.java]: `null` check `dirtiedAssets` ? Also, if `dirtiedAssets` is empty or null, just `return Optional.of(phase.getBlock(0));` ?

File: src/main/java/org/apache/mesos/scheduler/plan/Status.java [hunk 14]
```diff
@@ -4,9 +4,9 @@
  * Status of the block. Useful for reporting
  */
 public enum Status {
-  ERROR,
-  WAITING,
-  PENDING,
-  IN_PROGRESS,
-  COMPLETE
+  ERROR, // execution experienced an error
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/Status.java]: 👍  for updating documentation.

Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/Status.java]: Can you convert these to JavaDoc style comments `/** */` while you are at it ? That way these will become part of our JavaDocs.

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java [hunk 15]
```diff
@@ -25,93 +23,68 @@
...
-            TaskFailureListener failureListener,
             RecoveryRequirementProvider offerReqProvider,
             LaunchConstrainer launchConstrainer,
             FailureMonitor failureMonitor) {
         this.stateStore = stateStore;
         this.offerReqProvider = offerReqProvider;
         this.failureMonitor = failureMonitor;
         this.launchConstrainer = launchConstrainer;
-        this.failureListener = failureListener;
+        updatePlan(new ArrayList<>());
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: Nit: s/`new ArrayList<>()`/`Collections.EMPTY_LIST`/g (feel free to ignore this one)

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java [hunk 16]
```diff
@@ -203,9 +174,53 @@ public void forceComplete(UUID phaseId, UUID blockId) {
...
+                    Collection<Block> blocks = new ArrayList<>();
+
+                    // 2. remove any COMPLETED blocks for this task.
+                    for (Phase phase : this.plan.getPhases()) {
+                        blocks.addAll(phase.getBlocks().stream()
+                                .filter(block -> !(block.isComplete() && block.getName().equals(taskName)))
+                                .collect(Collectors.toList()));
+                    }
+
+                    // 3. Add new PENDING block.
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: 1. If a block was already pending, and for whatever reason (reconcile, etc.) we get another `TaskStatus` needing recovery, the code will add another `PENDING` block for it. I think we should make is explicit to only add a block in the phase if it's not already present.
2. Also, when a recovery Block is in `IN_PROGRESS` state, and the it gets another `TaskStatus`.

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java [hunk 17]
```diff
@@ -203,9 +174,53 @@ public void forceComplete(UUID phaseId, UUID blockId) {
...
+                    blocks.add(newBlock.get());
+
+                    updatePlan(blocks);
+                }
+            }
+        } catch (TaskException ex) {
+            LOGGER.error("Error updating status for task (id=%s)".format(status.getTaskId().toString()), ex);
```
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: Style nit:
`LOGGER.error("Error updating status for task (id={}). Reason: {}", status.getTaskId(), ex);`

File: src/test/java/org/apache/mesos/scheduler/DefaultSchedulerTest.java [hunk 18]
```diff
@@ -554,4 +515,59 @@ private int countBlocks(Plan plan) {
 
         return i;
     }
+
+    private Protos.TaskID installBlock(int phaseIndex, int blockIndex, Protos.Offer offer) {
```
Comment (mohitsoni) [inline src/test/java/org/apache/mesos/scheduler/DefaultSchedulerTest.java]: 👍  for creating helpers.

File: src/main/java/org/apache/mesos/curator/CuratorStateStore.java (hunk 2 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/curator/CuratorStateStore.java]: No, this is only used to serve the `/v1/state/properties/suppressed` endpoint

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java (hunk 3 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java]: Validate that it's non-null, or that it's non-empty?  We don't do null checks anywhere else in the code (and rightly so, it would be too verbose).  And it's valid for `dirtiedAssets` to be empty.

File: src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java (hunk 9 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java]: fixed

File: src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java (hunk 10 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/PlanCoordinator.java]: fixed

File: src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java (hunk 12 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/RandomRecoveryStrategy.java]: addressed above

File: src/main/java/org/apache/mesos/scheduler/plan/ReconciliationStrategy.java (hunk 13 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/ReconciliationStrategy.java]: addressed above, and the existing code has that same effect.

File: src/main/java/org/apache/mesos/scheduler/plan/Status.java (hunk 14 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/plan/Status.java]: fixed

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java (hunk 17 above)
Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: fixed

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java (hunk 1 above)
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: @mgummelt:  It's used here to mark Tasks as permanently failed correct? https://github.com/mesosphere/dcos-commons/blob/0df897784b8e68e42a31c09c9af79ea25ed3fc01/src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java#L275

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java (hunk 16 above)
Comment (gabrielhartmann) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: @mohitsoni:  This PR is for suppress revive right?  I think if we want to make improvements to the recovery manager's behavior we should probably do that in a different PR.  Do you want to make a JIRA to track the improvements you'd like?

/cc @mgummelt

Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: His comments are related to the correctness of my changes to this class, which I needed to make to support the observer pattern.  I'm addressing them now.

Comment (mgummelt) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: Fixed, and added regression test.

**Comment by mgummelt** [comment]
cc @mohitsoni please re-review

File: src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java (hunk 3 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/plan/DefaultInstallStrategy.java]: Yeah, it will work fine when `dirtiedAssets` is empty. 
I think not-doing null check in methods which are not part of public interfaces is okay. But, I think we should exercise caution by validating arguments of methods which are part of public interfaces like `PhaseStrategy`. If we are not doing it at other places in code, maybe we should start doing it.

File: src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java (hunk 16 above)
Comment (mohitsoni) [inline src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java]: @gabrielhartmann @mgummelt refactored this class to make use of `update()` method for receiving task updates, and regenerating the plan. Instead of the previous model, where the plan was re-generated only with each offer cycle. These comments are to address bug(s) introduced during that refactoring.

I think @mgummelt's approach is better is terms of consistency with PlanManager design, and it also allows for a single way for all PlanManagers to receives TaskStatuses via Observer pattern.

**Comment by mohitsoni** [comment]
👍
